File: JensonGodotKit/Sources/JensonGodotKit/JensonTimeline.swift
//
//  JensonTimeline.swift
//  Indexing Your Heart
//
//  Created by Marquis Kurt on 5/30/23.
//
//  This file is part of Indexing Your Heart.
//
//  Indexing Your Heart is non-violent software: you can use, redistribute, and/or modify it under the terms of the
//  CNPLv7+ as found in the LICENSE file in the source code root directory or at
//  <https://git.pixie.town/thufie/npl-builder>.
//
//  Indexing Your Heart comes with ABSOLUTELY NO WARRANTY, to the extent permitted by applicable law. See the CNPL for
//  details.

import JensonKit
import SwiftGodot

/// A Godot node that registers a Jenson timeline and attempts to update any children UI elements to reflect the current
/// event in the story.
@NativeHandleDiscarding
public class JensonTimeline: Node {
    static var timelineFinishedSignalName = StringName("timeline_finished")

    enum AnimationName: String {
        case startTimeline = "start_timeline"
        case speech

        var stringName: StringName {
            StringName(stringLiteral: rawValue)
        }
    }

    enum ChildPath: String {
        case animationPlayer = "AnimationPlayer"
        case background = "Background"
        case choiceMenu = "Choice Menu"
        case leftSpeaker = "Left Speaker"
        case rightSpeaker = "Right Speaker"
        case singleSpeaker = "Single Speaker"
    }

    enum ImageRefreshPriorityLayer: Int {
        case background = -1
        case speakerSingle = 0
        case speakerLeft = 1
        case speakerRight = 2
        case unknown = -999
    }

    /// Whether comment events should be displayed.
    @Autovariant public var displayCommentary: Bool = false

    /// The script being loaded into the timeline.
    @Autovariant public var script: String = "" {
        didSet {
            if !Engine.isEditorHint() {
                loadScript()
            }
        }
    }

    @SceneTree(path: ChildPath.animationPlayer.rawValue)
    private var animator: AnimationPlayer?

    @SceneTree(path: ChildPath.background.rawValue)
    private var backgroundLayer: TextureRect?
    private var choices = [String: [JensonEvent]]()
    private var choiceTemplate: Button?
    private var currentEvent: JensonEvent?
    private var finished = false
    private var initialized = false

    @SceneTree(path: ChildPath.choiceMenu.rawValue)
    private var menu: VBoxContainer?

    private var reader: JensonReader?

    @SceneTree(path: ChildPath.leftSpeaker.rawValue)
    private var speakerLeft: TextureRect?

    @SceneTree(path: ChildPath.rightSpeaker.rawValue)
    private var speakerRight: TextureRect?

    @SceneTree(path: ChildPath.singleSpeaker.rawValue)
    private var speakerSingle: TextureRect?
    private var timeline = [JensonEvent]()
    private var whoLabel: Label?
    private var whatLabel: Label?

    public required init() {
        JensonTimeline.initializeClass()
        super.init()
    }

    override public func _ready() {
        super._ready()
        whoLabel = findChild(pattern: "Who Label", recursive: true) as? Label
        whatLabel = findChild(pattern: "What Label", recursive: true) as? Label

        choiceTemplate = menu?.getChild(idx: 0) as? Button

        guard !Engine.isEditorHint() else { return }
        menu?.visible = false
        choiceTemplate?.visible = false
        whoLabel?.text = ""
        whatLabel?.text = ""

        if !displayCommentary {
            GD.print("Removing developer commentary nodes.")
            timeline.removeAll { event in
                event.type == .comment
            }
        }

        preloadRefreshedData()
        animator?.play(name: AnimationName.startTimeline.stringName)
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { [weak self] in
            self?.initialized = true
            self?.next()
        }
    }

    override public func _input(event _: InputEvent) {
        if Input.isActionPressed(action: .timelineNext) || Input.isMouseButtonPressed(button: .left) {
            handleNextEvent()
        }
    }

    func handleNextEvent() {
        guard let animator else {
            next()
            return
        }
        if let menu, menu.visible { return }
        if animator.isPlaying(), animator.currentAnimation != AnimationName.startTimeline.rawValue, let whatLabel {
            animator.stop()
            whatLabel.visibleRatio = 1
            return
        }
        next()
    }

    func loadScript() {
        do {
            let reader = try JensonReader(resource: script)
            guard let file = try reader?.decode() else {
                GD.pushError("Decoded file for \(String(describing: script)) returned nil.")
                return
            }
            timeline = file.timeline
            if timeline.isEmpty {
                GD.pushWarning("The file loaded, but the timeline is empty.")
                return
            }
        } catch {
            GD.pushError("Failed to load reader for script", error.localizedDescription)
        }
    }

    func next() {
        guard !timeline.isEmpty else {
            if !finished {
                finished.toggle()
                GD.print("Timeline has finished.")
                try? emitSignal(JensonTimeline.timelineFinishedSignalName)
            }
            GD.pushWarning("Attempted to move to an empty slot.")
            return
        }
        currentEvent = timeline.removeFirst()
        if let currentEvent {
            setup(event: currentEvent)
        } else {
            GD.print("Timeline has completed.")
            try? emitSignal(JensonTimeline.timelineFinishedSignalName)
        }
    }

    func preloadRefreshedData() {
        guard !Engine.isEditorHint(), let first = timeline.first else { return }
        if first.type != .refresh {
            GD.pushWarning("Next event doesn't require initial setup.")
            return
        }
        GD.print("Running initial refresh contents.")
        next()
        initialized = true
    }

    func refreshScene(using triggers: [JensonRefreshContent]) {
        triggers.forEach { trig in
            switch trig.kind {
            case .image:
                refreshImageLayer(with: trig)
            default:
                GD.pushWarning("Unsupported refresh kind: \(trig.kind.rawValue). This trigger will be skipped.")
            }
        }
        guard initialized else { return }
        next()
    }

    func refreshImageLayer(with trigger: JensonRefreshContent) {
        guard let priority = ImageRefreshPriorityLayer(rawValue: trigger.priority ?? -999) else {
            GD.pushWarning("Skipping priority-less image trigger.")
            return
        }
        switch priority {
        case .background:
            guard let tex: Texture2D = GD.load(path: "res://resources/backgrounds/\(trigger.resourceName).png") else {
                GD.pushWarning("Unable to find image resource \(trigger.resourceName).")
                return
            }
            backgroundLayer?.texture = tex
        case .speakerSingle:
            guard let tex: Texture2D = GD.load(path: "res://resources/characters/\(trigger.resourceName).png") else {
                GD.pushWarning("Unable to find image resource \(trigger.resourceName).")
                return
            }
            speakerSingle?.texture = tex
        case .speakerLeft:
            guard let tex: Texture2D = GD.load(path: "res://resources/characters/\(trigger.resourceName).png") else {
                GD.pushWarning("Unable to find image resource \(trigger.resourceName).")
                return
            }
            speakerLeft?.texture = tex
        case .speakerRight:
            guard let tex: Texture2D = GD.load(path: "res://resources/characters/\(trigger.resourceName).png") else {
                GD.pushWarning("Unable to find image resource \(trigger.resourceName).")
                return
            }
            speakerRight?.texture = tex
        case .unknown:
            GD.pushWarning("Skipping refresh with unknown priority.")
        }
    }

    func setup(event: JensonEvent) {
        switch event.type {
        case .dialogue:
            setupDialogue(with: event)
        case .question:
            setupDialogue(with: event)
            guard let question = event.question else {
                GD.pushWarning("Question event expected a question object, but received nil.")
                return
            }
            setupQuestion(question: question)
        case .refresh:
            if let refreshTriggers = event.refresh {
                refreshScene(using: refreshTriggers)
            }
        case .comment where displayCommentary:
            setupDialogue(with: event, overrideSpeaker: "Developer Commentary")
        default:
            GD.pushWarning("Unknown event type: \(event.type.rawValue). This will be skipped.")
            next()
        }
    }

    func setupDialogue(with event: JensonEvent, overrideSpeaker speaker: String? = nil) {
        whoLabel?.text = speaker ?? event.who
        whatLabel?.text = event.what
        animator?.play(name: AnimationName.speech.stringName, customSpeed: Double(event.what.count) / 4.0)
    }

    func setupQuestion(question: JensonQuestion) {
        guard let menu else { return }
        choices.removeAll()
        question.options.forEach { choice in
            choices[choice.name] = choice.events
        }
        menu.getChildren()
            .filter { child in child != choiceTemplate }
            .forEach(menu.removeChild(node:))
        question.options.map(\.name).forEach { choiceName in
            if let newButton = choiceTemplate?.duplicate() as? Button {
                newButton.text = choiceName
                _ = try? newButton.pressed.connect { [weak self] in
                    self?.menu?.visible = false
                    guard let events = self?.choices[choiceName] else {
                        GD.pushWarning("No option with key '\(choiceName)' exists. No new events will be inserted.")
                        return
                    }
                    self?.timeline.insert(contentsOf: events, at: 0)
                    self?.next()
                }
                newButton.visible = true
                menu.addChild(node: newButton)
            }
        }
        menu.visible = true
    }
}

File: Shounin/scripts/UI/JensonTimeline.cs
#region Copyright
//
//  JensonTimeline.cs
//  Indexing Your Heart
//
//  Created by Marquis Kurt on 21/7/2024.
//
//  This file is part of Indexing Your Heart.
//
//  Indexing Your Heart is non-violent software: you can use, redistribute, and/or modify it under the terms of the
//  CNPLv7+ as found in the LICENSE file in the source code root directory or at
//  <https://git.pixie.town/thufie/npl-builder>.
//
//  Indexing Your Heart comes with ABSOLUTELY NO WARRANTY, to the extent permitted by applicable law. See the CNPL for
//  details.
#endregion

using System.Collections.Generic;
using System.Linq;
using Godot;
using IndexingYourHeart.Utils;
using IndexingYourHeart.Utils.Animation;
using Jenson.NET;
using Jenson.NET.Models;
using System;

namespace IndexingYourHeart.UI
{
    /// <summary>
    /// A node that can display a Jenson script for dialogue or other story-based elements.
    /// </summary>
    public partial class JensonTimeline : Control
    {
        /// <summary>
        ///  A representation of the various states the current timeline can be in.
        /// </summary>
        public enum TimelineState
        {
            /// <summary>
            /// The initial state, where the node has been created, but the script hasn't been loaded yet.
            /// </summary>
            Initial,

            /// <summary>
            /// The script has been loaded, but the timeline hasn't started execution yet.
            /// </summary>
            Loaded,

            /// <summary>
            /// The script has been loaded and the timeline has started execution. This state is typically also used to
            /// indicate that the starting animation is playing, and any interactions should be ignored.
            /// </summary>
            Started,

            /// <summary>
            /// The script has been loaded and the timeline is actively playing events.
            /// </summary>
            Playing,

            /// <summary>
            /// The timeline has finished execution, and there are no more timeline events to read.
            /// </summary>
            Ended
        }

        public enum ImageRefreshPriorityLayer
        {
            Background = -1,
            SpeakerSingle = 0,
            SpeakerLeft = 1,
            SpeakerRight = 2
        }

        private static TimelineState[] UnsafeRefreshStates => [TimelineState.Initial, TimelineState.Loaded, TimelineState.Ended];

        /// <summary>
        /// The path to the timeline script to load. Ideally set in the editor, but can be manually instantiated.
        /// </summary>
        [Export(PropertyHint.File, "*.jenson")]
        public string Script = "";

        /// <summary>
        /// The current state of the timeline.
        /// </summary>
        public TimelineState CurrentTimelineState => _timelineState;

        private Dictionary<string, List<IJensonEvent>> choices = new();
        private Button choiceTemplate;
        private IJensonEvent currentEvent;
        private JensonReader reader;
        private List<IJensonEvent> timeline;
        private TimelineState _timelineState = TimelineState.Initial;

        #region Children Nodes
        private AnimationPlayer animator;
        private TextureRect backgroundLayer;
        private VBoxContainer menu;
        private TextureRect speakerLeft;
        private TextureRect speakerRight;
        private TextureRect speakerSingle;
        private Label whoLabel;
        private Label whatLabel;
        #endregion

        public override void _Ready()
        {
            animator = GetNode<AnimationPlayer>("AnimationPlayer");
            backgroundLayer = GetNode<TextureRect>("Background");
            menu = GetNode<VBoxContainer>("Choice Menu");
            speakerLeft = GetNode<TextureRect>("Multi Speakers/Left Speaker");
            speakerRight = GetNode<TextureRect>("Multi Speakers/Right Speaker");
            speakerSingle = GetNode<TextureRect>("Single Speaker");

            whoLabel = (Label)FindChild("Who Label", true);
            whatLabel = (Label)FindChild("What Label", true);

            choiceTemplate = menu.GetChild<Button>(0);

            menu.Visible = false;
            choiceTemplate.Visible = false;
            whoLabel.Text = "";
            whatLabel.Text = "";

            // Listen for when the animation finishes rather than using a dispatch queue.
            animator.AnimationFinished += delegate
            {
                if (_timelineState == TimelineState.Started)
                {
                    _timelineState = TimelineState.Playing;
                    Next();
                }
            };

            if (Script == string.Empty)
            {
                GD.PushWarning(
                    "Script was not set in the editor. Set the Script property and call LoadScript() to execute the timeline."
                );
                return;
            }
            LoadScript();
            StartTimeline();
        }

        public override void _Input(InputEvent @event)
        {
            if (!Visible)
                return;
            if (Input.IsActionPressed("timeline_next") || Input.IsMouseButtonPressed(MouseButton.Left))
                HandleNextEvent();
            base._Input(@event);
        }

        /// <summary>
        /// Loads the current script into the timeline node.
        /// </summary>
        public void LoadScript()
        {
            using var file = FileAccess.Open(Script, FileAccess.ModeFlags.Read);
            reader = new JensonReader(file.GetAsText());
            timeline = reader.Parse().timeline.ToList();
            _timelineState = TimelineState.Loaded;
            EmitSignal(SignalName.TimelineLoaded);
        }

        /// <summary>
        /// Start the playback of the current timeline.
        /// </summary>
        /// <remarks>
        /// This method assumes that the script has already been loaded via <see cref="LoadScript"/>.
        /// </remarks>
        public void StartTimeline()
        {
            if (_timelineState != TimelineState.Loaded) return;
            if (timeline.First().EventType == JensonEventType.Refresh)
                Next();

            _timelineState = TimelineState.Started;
            animator.Play("start_timeline");
        }

        private void HandleNextEvent()
        {
            if (animator == null || menu.Visible || _timelineState == TimelineState.Started)
                return;
            if (animator.IsPlaying() && animator.CurrentAnimation != "start_timeline")
            {
                SkipAnimation();
                return;
            }
            Next();
        }

        private void Next()
        {
            if (timeline.Count == 0)
            {
                HandleEmptyTimeline();
                return;
            }
            currentEvent = timeline.RemoveFirst();
            SetupWithCurrentEvent();
        }

        private void HandleEmptyTimeline()
        {
            if (_timelineState != TimelineState.Ended)
            {
                ResetTimelineState();
                _timelineState = TimelineState.Ended;
                GD.Print("Timeline has finished.");
                EmitSignal(SignalName.TimelineFinished);

                return;
            }
            GD.PushWarning("Attempted to move to an empty slot.");
        }

        private void ResetTimelineState()
        {
            whoLabel.Text = string.Empty;
            whatLabel.Text = string.Empty;

            List<TextureRect> rects = [backgroundLayer, speakerSingle, speakerLeft, speakerRight];
            foreach (var rect in rects)
            {
                this.Animate(UIAnimation.InterpolatingSpring(), new UIAnimationProperty
                {
                    Target = rect,
                    Property = "modulate",
                    EndState = Colors.Transparent
                }, () =>
                {
                    rect.Texture = null;
                    rect.Modulate = Colors.White;
                });
            }
            animator.Play("start_timeline", customSpeed: -1, fromEnd: true);
        }

        private void RefreshSceneWithCurrentEvent()
        {
            RefreshEvent refreshEvent = (RefreshEvent)currentEvent;
            switch (refreshEvent.Kind)
            {
                case "image":
                    RefreshImageWithCurrentEvent();
                    break;
                default:
                    GD.PushWarning($"Unsupported refresh kind: {refreshEvent.Kind}. This trigger will be skipped.");
                    break;
            }
            if (UnsafeRefreshStates.Contains(_timelineState))
                return;
            Next();
        }

        private void RefreshImageWithCurrentEvent()
        {
            RefreshEvent refreshEvent = (RefreshEvent)currentEvent;
            switch (refreshEvent.Priority)
            {
                case (int)ImageRefreshPriorityLayer.Background:
                    LoadTextureIntoImage(backgroundLayer, refreshEvent, "backgrounds");
                    break;
                case (int)ImageRefreshPriorityLayer.SpeakerSingle:
                    LoadTextureIntoImage(speakerSingle, refreshEvent, "characters");
                    break;
                case (int)ImageRefreshPriorityLayer.SpeakerLeft:
                    LoadTextureIntoImage(speakerLeft, refreshEvent, "characters");
                    speakerLeft.FlipH = true;
                    break;
                case (int)ImageRefreshPriorityLayer.SpeakerRight:
                    LoadTextureIntoImage(speakerRight, refreshEvent, "characters");
                    break;
                default:
                    GD.PushWarning($"Unrecognized priority: {refreshEvent.Priority}. Skipping.");
                    break;
            }
        }

        private void LoadTextureIntoImage(TextureRect imageRect, RefreshEvent refreshEvent, string domain)
        {
            this.Animate(UIAnimation.InterpolatingSpring(), new UIAnimationProperty
            {
                Target = imageRect,
                Property = "modulate",
                EndState = Colors.Transparent
            }, () =>
            {
                if (refreshEvent.What == string.Empty)
                {
                    GD.PushWarning("Refresh image name is empty, assuming to clear.");
                    imageRect.Texture = null;
                    this.Animate(UIAnimation.InterpolatingSpring(), new UIAnimationProperty
                    {
                        Target = imageRect,
                        Property = "modulate",
                        EndState = Colors.Transparent
                    });
                    return;
                }
                var speakerRightPath = $"res://resources/{domain}/{refreshEvent.What}.png";
                var speakerRightTexture = GD.Load<Texture2D>(speakerRightPath);
                imageRect.Texture = speakerRightTexture;
                this.Animate(UIAnimation.InterpolatingSpring(), new UIAnimationProperty
                {
                    Target = imageRect,
                    Property = "modulate",
                    EndState = Colors.White
                });
            });
        }

        private void SetupButton(Button button, string choiceName)
        {
            button.Text = choiceName.ToUpper();
            button.Visible = true;
            button.Pressed += delegate
            {
                menu.Visible = false;
                if (!choices.ContainsKey(choiceName))
                    return;
                var events = choices[choiceName];
                var mergedTimeline = events;
                foreach (var timelineEvent in timeline)
                {
                    mergedTimeline.Add(timelineEvent);
                }
                timeline = mergedTimeline;
                Next();
            };

            menu.AddChild(button);
        }

        private void SetupWithCurrentEvent()
        {
            animator.Stop();
            switch (currentEvent.EventType)
            {
                case JensonEventType.Narration:
                    SetupNarration();
                    break;
                case JensonEventType.Dialogue:
                    SetupDialogueWithCurrentEvent();
                    break;
                case JensonEventType.Refresh:
                    RefreshSceneWithCurrentEvent();
                    break;
                case JensonEventType.Question:
                    QuestionEvent question = (QuestionEvent)currentEvent;
                    DialogueEvent questionDialogue = new DialogueEvent(question.Who, question.What);
                    SetupDialogue(questionDialogue);
                    SetupQuestion();
                    break;
                default:
                    GD.PushWarning($"Unknown event type: {currentEvent.EventType}. Skipping.");
                    Next();
                    break;
            }
        }

        private void SetupDialogue(DialogueEvent dialogue)
        {
            whoLabel.Text = dialogue.Who;
            whatLabel.Text = dialogue.What;
            SkipImageModulation();
            animator.Play("speech", (double)dialogue.What.Length / 0.25);
        }

        private void SetupDialogueWithCurrentEvent()
        {
            DialogueEvent dialogue = (DialogueEvent)currentEvent;
            whoLabel.Text = dialogue.Who;
            whatLabel.Text = dialogue.What;
            whatLabel.VisibleRatio = 0.0f;
            SkipImageModulation();

            this.Animate(UIAnimation.LinearEaseInOut(CalculateCPS(dialogue.What)), new UIAnimationProperty
            {
                Target = whatLabel,
                Property = "visible_ratio",
                EndState = 1.0f
            }, () =>
            {
                EmitSignal(SignalName.TimelineDialogueFired, whoLabel.Text, whatLabel.Text);
            });
        }

        private void SetupNarration()
        {
            NarrationEvent narration = (NarrationEvent)currentEvent;
            whatLabel.Text = narration.What;
            whatLabel.VisibleRatio = 0.0f;
            whoLabel.Text = "";
            SkipImageModulation();

            this.Animate(UIAnimation.LinearEaseInOut(CalculateCPS(narration.What)), new UIAnimationProperty
            {
                Target = whatLabel,
                Property = "visible_ratio",
                EndState = 1.0f
            }, () =>
            {
                EmitSignal(SignalName.TimelineDialogueFired, "<#narration#>", whatLabel.Text);
            });
        }

        // ReSharper disable once InconsistentNaming
        private float CalculateCPS(string line)
        {
            return line.Length / 40.0f;
        }

        private void SetupQuestion()
        {
            QuestionEvent question = (QuestionEvent)currentEvent;
            choices.Clear();
            foreach (var choice in question.Choices)
            {
                choices[choice.What] = choice.Events.ToList();
            }

            foreach (var child in menu.GetChildren().Where((child) => child != choiceTemplate))
            {
                menu.RemoveChild(child);
            }

            foreach (var choiceName in question.Choices.Select((choice) => choice.What))
            {
                Button newButton = (Button)choiceTemplate.Duplicate();
                SetupButton(newButton, choiceName);
            }

            menu.Visible = true;
        }

        private void SkipAnimation()
        {
            animator.Stop();
            whoLabel.VisibleRatio = 1;
            whatLabel.VisibleRatio = 1;

            SkipImageModulation();
        }

        private void SkipImageModulation()
        {
            backgroundLayer.Modulate = Colors.White;
            speakerLeft.Modulate = Colors.White;
            speakerRight.Modulate = Colors.White;
            speakerSingle.Modulate = Colors.White;
        }

        [Signal]
        public delegate void TimelineLoadedEventHandler();

        [Signal]
        public delegate void TimelineDialogueFiredEventHandler(string Who, string What);

        [Signal]
        public delegate void TimelineFinishedEventHandler();
    }
}
